client/protocols/backup.py

```diff
@@ -0,0 +1,179 @@
+from client.protocols.common import models
+from client.utils import auth
+import client
+import datetime
+import json
+import logging
+import os
+import pickle
+import socket
+import urllib.error
+import urllib.request
+
+log = logging.getLogger(__name__)
+
+class BackupProtocol(models.Protocol):
+
+    TIMEOUT = 500
+    RETRY_LIMIT = 5
+    BACKUP_FILE = ".ok_messages"
+    SUBMISSION_ENDPOINT = '{prefix}://{server}/api/v1/submission?'
+
+    def run(self, messages):
+        if self.args.local:
+            return
+
+        message_list = self.load_unsent_messages()
+        message_list.append(messages)
+
+        self.check_ssl()
+
+        access_token = auth.authenticate(self.args.authenticate)
+        log.info('Authenticated with access token %s', access_token)
+
+        response = self.send_all_messages(access_token, message_list)
+
+        if isinstance(response, dict):
+            print('Backup successful for user: '
+                  '{0}'.format(response['data']['email']))
+            if self.args.submit or self.args.backup:
+                print('URL: https://ok-server.appspot.com/#/'
+                      '{0}/submission/{1}'.format(response['data']['course'],
+                                                  response['data']['key']))
+            if self.args.backup:
+                print('NOTE: this is only a backup. '
+                      'To submit your assignment, use:\n'
+                      '\tpython3 ok --submit')
+
+        self.dump_unsent_messages(message_list)
+
+
+    @classmethod
+    def load_unsent_messages(cls):
+        message_list = []
+        try:
+            with open(cls.BACKUP_FILE, 'rb') as fp:
+                message_list = pickle.load(fp)
+            log.info('Loaded %d backed up messages from %s',
+                     len(message_list), cls.BACKUP_FILE)
+        except (IOError, EOFError) as e:
+            log.info('Error reading from ' + cls.BACKUP_FILE + \
+                     ', assume nothing backed up')
+        return message_list
+
+
+    @classmethod
+    def dump_unsent_messages(cls, message_list):
+        with open(cls.BACKUP_FILE, 'wb') as f:
+            log.info('Save %d unsent messages to %s', len(message_list),
+                     cls.BACKUP_FILE)
+
+            pickle.dump(message_list, f)
+            os.fsync(f)
+
+
+    def check_ssl(self):
```

Comment:
I think checking whether ssl is available should happen in ok-main and each protocol just assume that ssl is actually available (and look at args.insecure of course). The reason for this is that other protocols (e.g. Export Protocol) will also be communicating with the server and will also need SSL.